pkg/endpointmanager/manager.go

```diff
@@ -27,145 +27,100 @@ import (
-	Mutex.Lock()
-	defer Mutex.Unlock()
+	case endpoint.ContainerIdPrefix:
+		return lookupDockerID(eid), nil
 
-	return LookupDockerIDLocked(id)
-}
+	case endpoint.DockerEndpointPrefix:
+		return lookupDockerEndpoint(eid), nil
 
-func lookupIPv4Locked(ipv4 string) *endpoint.Endpoint {
-	if ep, ok := endpointsAux[endpoint.NewID(endpoint.IPv4Prefix, ipv4)]; ok {
-		return ep
-	}
-	return nil
-}
+	case endpoint.ContainerNamePrefix:
+		return lookupDockerContainerName(eid), nil
 
-// LookupIPv4 looks up endpoint by IPv4 address
-func LookupIPv4(ipv4 string) *endpoint.Endpoint {
-	Mutex.Lock()
-	defer Mutex.Unlock()
+	case endpoint.PodNamePrefix:
+		return lookupPodNameLocked(eid), nil
 
-	return lookupIPv4Locked(ipv4)
-}
+	case endpoint.IPv4Prefix:
+		return lookupIPv4(eid), nil
 
-// LookupDockerIDLocked is identical to LookupDockerID() but with
-// endpointmanager.Mutex already held
-func LookupDockerIDLocked(id string) *endpoint.Endpoint {
-	if ep, ok := endpointsAux[endpoint.NewID(endpoint.ContainerIdPrefix, id)]; ok {
-		return ep
+	default:
+		return nil, fmt.Errorf("Unknown endpoint prefix %s", prefix)
 	}
-	return nil
 }
 
-func linkContainerID(ep *endpoint.Endpoint) {
-	endpointsAux[endpoint.NewID(endpoint.ContainerIdPrefix, ep.DockerID)] = ep
+// LookupCiliumID looks up endpoint by endpoint ID
+func LookupCiliumID(id uint16) *endpoint.Endpoint {
+	mutex.RLock()
+	ep := lookupCiliumID(id)
+	mutex.RUnlock()
+	return ep
 }
 
-// LinkContainerID links an endpoint and makes it searchable by Docker ID.
-// Mutex must be held
-func LinkContainerID(ep *endpoint.Endpoint) {
-	linkContainerID(ep)
+// LookupDockerID looks up endpoint by Docker ID
+func LookupDockerID(id string) *endpoint.Endpoint {
+	mutex.RLock()
+	ep := lookupDockerID(id)
+	mutex.RUnlock()
+	return ep
 }
 
-// UpdateReferences updates the mappings of various values to their corresponding
-// endpoints, such as DockerID, Docker Container Name, Pod Name, etc.
-func updateReferences(ep *endpoint.Endpoint) {
-	if ep.DockerID != "" {
-		linkContainerID(ep)
-	}
-
-	if ep.DockerEndpointID != "" {
-		endpointsAux[endpoint.NewID(endpoint.DockerEndpointPrefix, ep.DockerEndpointID)] = ep
-	}
-
-	if ep.IPv4.String() != "" {
-		endpointsAux[endpoint.NewID(endpoint.IPv4Prefix, ep.IPv4.String())] = ep
-	}
-
-	if ep.ContainerName != "" {
-		endpointsAux[endpoint.NewID(endpoint.ContainerNamePrefix, ep.ContainerName)] = ep
-	}
-
-	if ep.PodName != "" {
-		endpointsAux[endpoint.NewID(endpoint.PodNamePrefix, ep.PodName)] = ep
-	}
+// LookupIPv4 looks up endpoint by IPv4 address
+func LookupIPv4(ipv4 string) *endpoint.Endpoint {
+	mutex.RLock()
+	ep := lookupIPv4(ipv4)
+	mutex.RUnlock()
+	return ep
 }
 
 // UpdateReferences makes an endpoint available by all possible reference
 // fields as available for this endpoint (containerID, IPv4 address, ...)
 func UpdateReferences(ep *endpoint.Endpoint) {
-	Mutex.Lock()
-	defer Mutex.Unlock()
-
+	mutex.Lock()
 	updateReferences(ep)
+	mutex.Unlock()
```

Comment:
We seem to be mixing using defer and not. Is there a logic to it?

Reply:
defer has some overhead when used on small functions... so on this case it doesn't matter since `updateReferences` is called. I'll revert it back with defer